pkgs/applications/networking/sync/celeste/default.nix

```diff
@@ -0,0 +1,104 @@
+{ lib
+, stdenv
+, rust
+, rustPlatform
+, fetchFromGitHub
+, substituteAll
+, fetchpatch
+, pkg-config
+, wrapGAppsHook4
+, cairo
+, gdk-pixbuf
+, glib
+, graphene
+, gtk3
+, gtk4
+, libadwaita
+, libappindicator
+, libclang
+, librclone
+, pango
+}:
+
+let
+  # https://github.com/trevyn/librclone/pull/8
+  librclone-mismatched-types-patch = fetchpatch {
+    name = "use-c_char-to-be-platform-independent.patch";
+    url = "https://github.com/trevyn/librclone/commit/91fdf3fa5f5eea0dfd06981ba72e09034974fdad.patch";
+    hash = "sha256-8YDyUNP/ISP5jCliT6UCxZ89fdRFud+6u6P29XdPy58=";
+  };
+in rustPlatform.buildRustPackage rec {
+  pname = "celeste";
+  version = "0.4.0";
+
+  src = fetchFromGitHub {
+    owner = "hwittenborn";
+    repo = "celeste";
+    rev = "v${version}";
+    hash = "sha256-6DUoqxVmj6TgnSLvyXNmmb2os/UtIO4O+f4NaG0xEWg=";
+  };
+
+  cargoHash = "sha256-3RtKr8ifDI3R1xZWqSH/ty1GOWN0b+IE2icDekVTW4A=";
+
+  patches = [
+    (substituteAll {
+      src = ./target-dir.patch;
+      rustTarget = rust.toRustTarget stdenv.hostPlatform;
+    })
+  ];
+
+  postPatch = ''
+    pushd $cargoDepsCopy/librclone-sys
+    oldHash=$(sha256sum build.rs | cut -d " " -f 1)
+    patch -p2 < ${./librclone-path.patch}
+    substituteInPlace build.rs \
+      --subst-var-by librclone ${librclone}
+    substituteInPlace .cargo-checksum.json \
+      --replace $oldHash $(sha256sum build.rs | cut -d " " -f 1)
+    popd
+    pushd $cargoDepsCopy/librclone
+    oldHash=$(sha256sum src/lib.rs | cut -d " " -f 1)
+    patch -p1 < ${librclone-mismatched-types-patch}
+    substituteInPlace .cargo-checksum.json \
+      --replace $oldHash $(sha256sum src/lib.rs | cut -d " " -f 1)
+    popd
+  '';
+
+  nativeBuildInputs = [
+    pkg-config
+    rustPlatform.bindgenHook
+    wrapGAppsHook4
+  ];
+
+  # We need to build celeste-tray first because celeste/src/launch.rs reads that file at build time.
+  # Upstream does the same: https://github.com/hwittenborn/celeste/blob/765dfa2/justfile#L1-L3
+  cargoBuildFlags = [ "--bin" "celeste-tray" ];
+  postConfigure = ''
+    cargoBuildHook
+    cargoBuildFlags=
+  '';
+
+  RUSTC_BOOTSTRAP = 1;
```

Comment:
the sorting here looks a little weird, I would personally sort it this way

- nativeBuildInputs
- buildInputs
- postPatch or cargoBuildFlags
- postConfigure
- RUSTC_BOOTSTRAP

Reply:
Wouldn't `postPatch` idiomatically go after `src` instead of the inputs?

Reply:
[nixpkgs-hammering](https://github.com/jtojnar/nixpkgs-hammering/blob/main/lib/derivation-attributes.nix) seems to put `postPatch` after inputs, I also prefer grouping the inputs and phases separately.
Usually I would put `cargoBuildFlags` before phases but here there is an argument for putting it right before `postConfigure`.

Reply:
Yay, subjectivity. 😕 something something style guide something something.

@dotlambda I was going to take a final look at this tomorrow, though if you're fine with just @figsoda's review, feel free to merge.

Reply:
This is absolutely subjective, and that's why I approved this pull request. This is just a suggestion and feel free to not apply the change.